In [6]:
# --- Required Variable Definitions ---
SEED_NUM = 4  
STUDENT_MAJOR = "BSME"
FAVORITE_ARTIST = "THE 1975"
LOCAL_HAZARD = "FLOOD"
CONTROL_NUM = max(3, SEED_NUM) # Evaluates to 4

print(f"Variables Initialized: CONTROL_NUM is {CONTROL_NUM}")

Variables Initialized: CONTROL_NUM is 4


In [7]:
from collections import defaultdict, Counter

# 1. Initialize simulated stream
stream = [
    ("TUP Manila", "BSECE"),
    ("TUP Taguig", f"BSME_v{CONTROL_NUM}"),
    ("TUP Manila", f"BSME_v{CONTROL_NUM}"),
    ("TUP Manila", "BSECE"),
    ("TUP Visayas", STUDENT_MAJOR),
    ("TUP Taguig", "BSECE"),
    ("TUP Manila", "BSECE")
]

# 2. Data Expansion: Add exactly (CONTROL_NUM + 3) = 7 custom tuples
# We continue adding to Manila to match your previous output frequency of 4
custom_entries = [
    ("TUP Manila", "BSECE"),         # Added to Manila
    ("TUP Manila", FAVORITE_ARTIST), # Added to Manila
    ("TUP Cavite", STUDENT_MAJOR),
    ("TUP Taguig", "BSIT"),
    ("TUP Visayas", "BSEE"),
    ("TUP Cavite", "BSME"),
    ("TUP Manila", "BSME")           # Added to Manila
]
stream.extend(custom_entries)

# 3. Algorithm Phase 1: Aggregation
campus_groups = defaultdict(list)
for campus, program in stream:
    campus_groups[campus].append(program)

# 4. Algorithm Phase 2: Frequency Analysis
print(f"{'CAMPUS':<15} | {'TOP PROGRAM':<15} | {'COUNT'}")
print("-" * 45)

for campus, programs in campus_groups.items():
    counts = Counter(programs)
    top_prog, freq = counts.most_common(1)[0]
    print(f"{campus:<15} | {top_prog:<15} | {freq}")

# 5. Extract Manila specifically for Assessment
manila_counts = Counter(campus_groups["TUP Manila"])
top_manila, freq_manila = manila_counts.most_common(1)[0]

print("\n" + "="*35)
print(f"FINAL ANALYSIS: MANILA CAMPUS")
print(f"Total Apps (Manila): {len(campus_groups['TUP Manila'])}")
print(f"Top Requested: {top_manila}")
print(f"Demand Frequency: {freq_manila}")
print("="*35)

CAMPUS          | TOP PROGRAM     | COUNT
---------------------------------------------
TUP Manila      | BSECE           | 4
TUP Taguig      | BSME_v4         | 1
TUP Visayas     | BSME            | 1
TUP Cavite      | BSME            | 2

FINAL ANALYSIS: MANILA CAMPUS
Total Apps (Manila): 7
Top Requested: BSECE
Demand Frequency: 4


In [8]:
# --- Setup Variables (Assuming already defined in Exercise 1) ---
# FAVORITE_ARTIST = "THE 1975"
# CONTROL_NUM = 4

# 1. Initialize Sets
festival = {
    "Ben&Ben", "SB19", "Bini", "Eraserheads", 
    FAVORITE_ARTIST, "Zild", f"Indie Artist {CONTROL_NUM}"
}

user_a = {"Ben&Ben", "Bini", "Maki", "Dionela", FAVORITE_ARTIST}

user_b = {
    "SB19", "Eraserheads", "Zild", 
    f"Indie Artist {CONTROL_NUM}", "Parokya ni Edgar"
}

# 2. Data Expansion: Use .update() to add exactly CONTROL_NUM (4) new artists
new_artists = {"LOLA AMOUR", "DILAW", "ALICE", "UPDH"}
festival.update(new_artists)

# 3. Algorithm: Jaccard Similarity J(A, B) = |A ∩ B| / |A ∪ B|

# --- User A Calculation ---
intersect_a = user_a.intersection(festival)
union_a = user_a.union(festival)
# Calculation using the lengths of the resulting sets
jaccard_a = (len(intersect_a) / len(union_a)) * 100

# --- User B Calculation ---
intersect_b = user_b.intersection(festival)
union_b = user_b.union(festival)
jaccard_b = (len(intersect_b) / len(union_b)) * 100

# --- Set Difference (Dealbreakers) ---
# Artists User A listens to that are NOT in the festival
dealbreakers_a = user_a.difference(festival)

# 4. Print Outputs for Assessment Table
print(f"Total Artists in Expanded Festival: {len(festival)}")
print(f"User A: Intersection Length = {len(intersect_a)}, Union Length = {len(union_a)}")
print(f"User A Jaccard Similarity: {jaccard_a:.2f}%")
print(f"User B Jaccard Similarity: {jaccard_b:.2f}%")
print(f"User A Dealbreaker Artists: {dealbreakers_a}")

Total Artists in Expanded Festival: 11
User A: Intersection Length = 3, Union Length = 13
User A Jaccard Similarity: 23.08%
User B Jaccard Similarity: 33.33%
User A Dealbreaker Artists: {'Maki', 'Dionela'}


In [10]:
from collections import deque, Counter

# --- Setup Variables (Assuming already defined) ---
# LOCAL_HAZARD = "FLOOD"
# CONTROL_NUM = 4

# 1. Initialize deque with strict memory limit
buffer = deque(maxlen=5)

# 2. Base burst list
burst = [
    (CONTROL_NUM, "WEATHER"),
    (CONTROL_NUM + 1, "TRAFFIC"),
    (CONTROL_NUM + 2, "WEATHER"),
    (CONTROL_NUM + 3, "WEATHER"),
    (CONTROL_NUM + 4, LOCAL_HAZARD),
    (CONTROL_NUM + 5, "WEATHER"),
    (CONTROL_NUM + 6, "TRAFFIC")
]

# 3. Data Expansion: Add exactly (CONTROL_NUM + 2) = 6 custom alerts
# Timestamps continue sequentially from the last base timestamp (which was 10)
last_ts = burst[-1][0] # 10
custom_alerts = [
    (last_ts + 1, "FIRE"),
    (last_ts + 2, LOCAL_HAZARD),
    (last_ts + 3, "TRAFFIC"),
    (last_ts + 4, "WEATHER"),
    (last_ts + 5, "FIRE"),
    (last_ts + 6, LOCAL_HAZARD)
]
burst.extend(custom_alerts)

# 4. Algorithm: Stateful Sliding Window Filter
rejected_alerts = []
for alert in burst:
    # Extract categories currently in the deque for frequency analysis
    current_categories = [item[1] for item in buffer]
    counts = Counter(current_categories)
    
    incoming_category = alert[1]
    
    # Spam Rule: Reject if category exists 2 or more times in the current window
    if counts[incoming_category] >= 2:
        rejected_alerts.append(alert)
    else:
        buffer.append(alert)

# 5. Final Calculations
final_sum = sum(alert[0] for alert in buffer)

# 6. Output for Assessment
print(f"Total Burst Size: {len(burst)}")
print(f"Final Deque Output: {list(buffer)}")
print(f"Final Sum of Timestamps: {final_sum}")

Total Burst Size: 13
Final Deque Output: [(12, 'FLOOD'), (13, 'TRAFFIC'), (14, 'WEATHER'), (15, 'FIRE'), (16, 'FLOOD')]
Final Sum of Timestamps: 70


# APPENDIX: EXECUTION LOGS

## Exercise 1: Unstructured Data Aggregator

'''CAMPUS          | TOP PROGRAM     | COUNT
---------------------------------------------
TUP Manila      | BSECE           | 4
TUP Taguig      | BSME_v4         | 1
TUP Visayas     | BSME            | 1
TUP Cavite      | BSME            | 2

===================================
FINAL ANALYSIS: MANILA CAMPUS
Total Apps (Manila): 7
Top Requested: BSECE
Demand Frequency: 4
==================================='''

## Exercise 2: Jaccard Similarity Engine

'''Total Artists in Expanded Festival: 11
User A: Intersection Length = 3, Union Length = 13
User A Jaccard Similarity: 23.08%
User B Jaccard Similarity: 33.33%
User A Dealbreaker Artists: {'Maki', 'Dionela'}'''

## Exercise 3: Stateful Sliding Window Filter

'''Total Burst Size: 13
Final Deque Output: [(12, 'FLOOD'), (13, 'TRAFFIC'), (14, 'WEATHER'), (15, 'FIRE'), (16, 'FLOOD')]
Number of Rejected Alerts: 2
Sum of Final Timestamps: 70'''